primitives/rpc/src/tracing.rs

```diff
@@ -0,0 +1,98 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Types for working with tracing data
+
+use serde::{Serialize, Deserialize};
+
+use rustc_hash::FxHashMap;
+
+/// Container for all related spans and events for the block being traced.
+#[derive(Serialize, Deserialize, Clone, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct BlockTrace {
+	/// Hash of the block being traced
+	pub block_hash: String,
```

Comment:
why not H256?

Reply:
Won't it need to be allocated anyways since we always serialize?

Reply:
yeah it would allocate when serializing/deserializing but current we do this:

```rust
		let block_trace = BlockTrace {
			block_hash: block_id_as_string(id),
			parent_hash: block_id_as_string(parent_id),
			tracing_targets: targets.to_string(),
			storage_keys: storage_keys.to_string(),
			spans,
			events,
		};

```

where block_id is to H256 which already implements serde AFAIK so it unnecessary to convert it a String and then serialize it :)

Reply:
`block_id_as_string` has a special way of displaying that is a bit easier to read for the RPC caller:
```
fn block_id_as_string<T: BlockT>(block_id: BlockId<T>) -> String {
	 match block_id {
		BlockId::Hash(h) => HexDisplay::from(&h.encode()).to_string(),
		BlockId::Number(n) =>  HexDisplay::from(&n.encode()).to_string()
	}
}
```
So I think we would need to write a custom `Serialize` impl that called `block_id_as_string` .. I might give it a try if I have time.